### PR TITLE
sink/kafka: actually use the configured tx timeout

### PR DESCRIPTION
**Pull request by petrosagg**
The code correctly set the desired transaction timeout value when instantiating the producer but then incorrectl used the socket timeout as the limit for the commit call. This led the `librdkafka` client to stop trying after 60 seconds, which is the default value of the socket timeout.

This PR instead sets the timeout to `Timeout::Never` which according to the `librdkafka` documentation automatically sets the timeout to the remaining transaction time based on the transaction_timeout_ms setting.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Review by antiguru — APPROVED** [review]
Ohhh, good find.

**Review by benesch — COMMENTED** [review]
I don't follow this fix! Why is the commit operation itself taking 60s? That seems quite surprising. Is it that we were able to fire off millions of messages and it's taking Kafka (or librdkafka, I guess) more than 60s to work through the backlog?

In any case, I don't think using the `transaction_timeout` here is quite right! At least, the docs for librdkafka suggest passing -1 to block for as much time is remaining in the transaction timeout:

https://github.com/confluentinc/librdkafka/blob/cb8c19c43011b66c4b08b25e5150455a247e1ff3/src/rdkafka.h#L10464-L10474

Not sure how to accomplish that with the Rust bindings. One second...

File: src/storage/src/sink/kafka.rs [hunk 1]
```diff
@@ -499,7 +502,7 @@ impl TransactionalProducer {
 
         fail::fail_point!("kafka_sink_commit_transaction");
 
-        let timeout = self.socket_timeout;
+        let timeout = self.transaction_timeout;
         match self
             .spawn_blocking(move |p| p.commit_transaction(timeout))
```
Comment (benesch) [inline src/storage/src/sink/kafka.rs:507]:
```suggestion
        // In this context, `None` means "as much time is remaining before the transaction
        // times out", which is exactly the behavior that we want.
        let timeout = None;
        match self
            .spawn_blocking(move |p| p.commit_transaction(timeout))
```

**Review by benesch — CHANGES_REQUESTED** [review]
Requesting changes to hold the automerge.

**Comment by benesch** [comment]
> Not sure how to accomplish that with the Rust bindings. One second...

Looks like `None` does what we want here:

```rust
/// Specifies a timeout for a Kafka operation.
#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
pub enum Timeout {
    /// Time out after the specified duration elapses.
    After(Duration),
    /// Block forever.
    Never,
}

impl Timeout {
    /// Converts a timeout to Kafka's expected representation.
    pub(crate) fn as_millis(&self) -> i32 {
        match self {
            Timeout::After(d) => d.as_millis() as i32,
            Timeout::Never => -1,
        }
    }
}
```

**Comment by petrosagg** [comment]
> Is it that we were able to fire off millions of messages and it's taking Kafka (or librdkafka, I guess) more than 60s to work through the backlog?

Exactly, that call waits for all outstanding messages to be delivered and that might take time.

> In any case, I don't think using the `transaction_timeout` here is quite right! At least, the docs for librdkafka suggest passing -1 to block for as much time is remaining in the transaction timeout

Ah, great find! The `transaction_timeout` should only slightly be over the actual limit since after the tx starts we call `.send()` for each record as fast as possible and those calls never block (they are only adding things to a queue). 

Changing the PR to pass `Timeout::Never` here

**Comment by petrosagg** [comment]
Hmm, @benesch the code doesn't seem to agree with the code. What I see here is that if you pass `-1` it will set it to `transaction_timeout * 2` https://github.com/confluentinc/librdkafka/blob/master/src/rdkafka_txnmgr.c#L1336-L1344

**Comment by benesch** [comment]
That's the code for init_transactions? The code for commit transaction

https://github.com/confluentinc/librdkafka/blob/cb8c19c43011b66c4b08b25e5150455a247e1ff3/src/rdkafka_txnmgr.c#L2512-L2529

calls `rd_kafka_txn_curr_api_begin`, which seems to behave as desired?

https://github.com/confluentinc/librdkafka/blob/cb8c19c43011b66c4b08b25e5150455a247e1ff3/src/rdkafka_txnmgr.c#L552-L565

**Comment by petrosagg** [comment]
Indeed, but it turns out their code still contradicts their docs since that function simply sets the timeout to the `transaction_timeout` configuration without tracking remaining time: https://github.com/confluentinc/librdkafka/blob/cb8c19c43011b66c4b08b25e5150455a247e1ff3/src/rdkafka_txnmgr.c#L276-L292

In any case, I agree we should make our calling code use `Timeout::Never` here and if they ever add tx remaining time tracking (and we manage to upgrade librdkafka :rip:) we get it for free.

**Review by benesch — APPROVED** [review]

**Comment by benesch** [comment]
> Indeed, but it turns out their code still contradicts their docs since that function simply sets the timeout to the `transaction_timeout` configuration without tracking remaining time: https://github.com/confluentinc/librdkafka/blob/cb8c19c43011b66c4b08b25e5150455a247e1ff3/src/rdkafka_txnmgr.c#L276-L292

Indeed! 😠 

> In any case, I agree we should make our calling code use `Timeout::Never` here and if they ever add tx remaining time tracking (and we manage to upgrade librdkafka :rip:) we get it for free.

Sounds like a plan to me.

**Comment by benesch** [comment]
@petrosagg the SSH connection test looks to be reliably failing. Not sure what's up there, but seems definitely related to this PR. :/

**Comment by petrosagg** [comment]
Looking into it, I think we fail to parse a specific error from librdkafka and we were previously relying on the timeout

**Comment by petrosagg** [comment]
Recording the new findings also discussed in Slack:

I think `librdkafka` is not honoring their docs in yet another way. The first thing the commit function does is to call this begin function https://github.com/confluentinc/librdkafka/blob/master/src/rdkafka_txnmgr.c#L2526-L2528

Which sets `abs_timeout` based on the value of `timeout_ms` and `rd_false`. The false parameter tells it to NOT cap the absolute timeout value. And so `abs_timeout` remains "infinite" and then proceeds to call `rd_kafka_txn_op_req` which then gets retried forever and our test fails.

I will revert the PR to the original version since that seems to be the correct way to use this API.

**Review by benesch — APPROVED** [review]

**Comment by benesch** [comment]
What a mess. Thanks, @petrosagg.

**Pull request merged**